src/api/core/tokenManager.ts

```diff
@@ -120,97 +124,130 @@ export class TokenManager {
     return this.initPromise;
   }
 
+  /** Terminate token manager and clear all tokens
+   */
   public terminate(): void {
     this.storage.clear();
-    clearInterval(this.refreshInterval);
+    this.refreshes.forEach((interval) => clearInterval(interval));
+    this.refreshes = [];
   }
 
+  /**
+   * Set specific token to use by default
+   * @param {string} auth authentication alias
+   */
   public setDefault(auth: string): void {
     this.storage.setDefault(auth);
   }
 
+  /**
+   * Switch back to apikey token
+   */
   public resetDefault(): void {
     this.storage.setDefault(APIKEY_DEFAULT_ALIAS);
   }
 
+  /**
+   * Add new token pair and run refresh digest
+   * @param {string} auth Authenticate alias
+   * @param {Tokens} tokens Token pair
+   * @param {boolean} defaults Whether to use this token by default
+   */
   public addTokens(auth: string, tokens: Tokens, defaults?: boolean) {
     this.storage.setTokens(auth, tokens, defaults);
-    this.refreshToken(auth);
+    this.startRefreshDigest(auth);
   }
 
-  private refreshToken(auth: string) {
-    this.refreshInterval = setInterval(() => {
-      this.logger.debug("tokenManager", `refresh ${auth} token`);
-      this.refresh(auth)
-        .catch((err: Error) => {
-          this.logger.error("tokenManager", err.message);
-          this.terminate();
-        });
-    }, DELAY);
+  /**
+   * Start refreshing digest for the specific auth
+   * @ignore
+   */
+  private startRefreshDigest(auth: string) {
+    this.refreshes.push(
+      setInterval(() => {
+        this.logger.debug("tokenManager", `refresh ${auth} token`);
+        this.refresh(auth)
+          .catch(() => this.terminate());
+      }, DELAY)
+    );
   }
 
+  /**
+   * Login to the project using APK method
+   * @ignore
+   */
   private login({auth = APIKEY_DEFAULT_ALIAS, key, secret}: IAuthOptions): Promise<this> {
-
-    let defers: any;
-    this.defers[auth] = new Promise((resolve, reject) => defers = { resolve, reject });
-
-    return this.requestAdapter
-      .execute(this.authUrl, {
-        body: {
-          method: "apk",
-          key,
-          secret,
-        },
-        method: Methods.POST,
-      })
-      .then((tokens: Tokens) => {
+    return this.obtainTokens(auth, this.authUrl, { method: "apk", key, secret })
+      .then((tokens) => {
         this.addTokens(auth, tokens, true);
-        defers.resolve(tokens.access_token);
         return this;
-      })
-      .catch((err: Error) => {
-        defers.reject(err);
-        this.logger.error("tokenManager", err.message);
-        throw new Error(`Unable to authenticate: ${err.message}`);
       });
   }
 
-  private refresh(auth: string): Promise<any> {
+  /**
+   * Refresh the token
+   * @ignore
+   */
+  private refresh(auth: string): Promise<this> {
 
     const tokens = this.storage.getTokens(auth);
 
     if (!tokens || !tokens.refresh_token) {
       return Promise.reject(`There is no refresh token for ${auth}`);
```

Comment:
it should reject an `Error` object instead of a string

Reply:
👍